File: src/main/com/intellij/lang/jsgraphql/ide/validation/GraphQLGeneralErrorFilter.java
package com.intellij.lang.jsgraphql.ide.validation;

import com.intellij.lang.jsgraphql.psi.*;
import com.intellij.lang.jsgraphql.psi.impl.GraphQLDirectivesAware;
import com.intellij.lang.jsgraphql.types.GraphQLError;
import com.intellij.lang.jsgraphql.types.validation.ValidationError;
import com.intellij.lang.jsgraphql.types.validation.ValidationErrorType;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class GraphQLGeneralErrorFilter implements GraphQLErrorFilter {
    @Override
    public boolean isIgnored(@NotNull Project project,
                             @Nullable GraphQLError error,
                             @Nullable PsiElement element) {
        if (!(error instanceof ValidationError)) {
            return false;
        }

        ValidationErrorType errorType = ((ValidationError) error).getValidationErrorType();
        if (isInsideTemplateElement(element)) {
            // error due to template placeholder replacement, so we can ignore it for '___' replacement variables
            if (errorType == ValidationErrorType.UndefinedVariable) {
                return true;
            }
        }

        if (errorType == ValidationErrorType.SubSelectionRequired) {
            // apollo client 2.5 doesn't require sub selections for client fields
            final GraphQLDirectivesAware directivesAware = PsiTreeUtil.getParentOfType(element, GraphQLDirectivesAware.class);
            if (directivesAware != null) {
                boolean ignoreError = false;
                for (GraphQLDirective directive : directivesAware.getDirectives()) {
                    if ("client".equals(directive.getName())) {
                        ignoreError = true;
                    }
                }
                if (ignoreError) {
                    return true;
                }
            }
        }

        if (errorType == ValidationErrorType.MisplacedDirective) {
            // graphql-java KnownDirectives rule only recognizes executable directive locations, so ignore
            // the error if we're inside a type definition
            return PsiTreeUtil.getParentOfType(element, GraphQLTypeSystemDefinition.class) != null;
        }
        return false;
    }

    boolean isInsideTemplateElement(@Nullable PsiElement psiElement) {
        return PsiTreeUtil.findFirstParent(
            psiElement, false,
            el -> el instanceof GraphQLTemplateDefinition || el instanceof GraphQLTemplateSelection || el instanceof GraphQLTemplateVariable
        ) != null;
    }

}

File: src/main/com/intellij/lang/jsgraphql/ide/validation/GraphQLGeneralErrorFilter.kt
package com.intellij.lang.jsgraphql.ide.validation

import com.intellij.lang.jsgraphql.ide.injection.GraphQLInjectionUtils
import com.intellij.lang.jsgraphql.psi.GraphQLIdentifier
import com.intellij.lang.jsgraphql.psi.GraphQLTypeSystemDefinition
import com.intellij.lang.jsgraphql.schema.GraphQLKnownTypes
import com.intellij.lang.jsgraphql.schema.library.GraphQLLibraryManager
import com.intellij.lang.jsgraphql.types.GraphQLError
import com.intellij.lang.jsgraphql.types.language.NamedNode
import com.intellij.lang.jsgraphql.types.schema.idl.errors.DirectiveIllegalArgumentTypeError
import com.intellij.lang.jsgraphql.types.schema.idl.errors.QueryOperationMissingError
import com.intellij.lang.jsgraphql.types.schema.idl.errors.TypeExtensionMissingBaseTypeError
import com.intellij.lang.jsgraphql.types.validation.ValidationError
import com.intellij.lang.jsgraphql.types.validation.ValidationErrorType
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.intellij.psi.impl.source.tree.injected.InjectedLanguageUtil
import com.intellij.psi.util.PsiTreeUtil

class GraphQLGeneralErrorFilter : GraphQLErrorFilter {

  override fun isGraphQLErrorSuppressed(
    project: Project,
    error: GraphQLError,
    element: PsiElement?,
  ): Boolean {
    if (error.message.contains(GraphQLInjectionUtils.GRAPHQL_EXTERNAL_FRAGMENT)) {
      return true
    }

    val isInInjectedLanguageBoundary = intersectsInjectedFragments(element)
    if (isInInjectedLanguageBoundary) {
      if ((error is ValidationError && error.validationErrorType == ValidationErrorType.WrongType) ||
          error is DirectiveIllegalArgumentTypeError) {
        return true
      }
    }

    val namedNode = error.node as? NamedNode<*>

    if (error is QueryOperationMissingError) {
      return true
    }

    if (error is TypeExtensionMissingBaseTypeError && namedNode?.name == GraphQLKnownTypes.QUERY_TYPE) {
      val roots = GraphQLLibraryManager.getInstance(project).libraryRoots.map { it.path }.toSet()
      return namedNode.sourceLocation?.sourceName in roots
    }

    if (error is ValidationError && error.validationErrorType == ValidationErrorType.MisplacedDirective) {
      // graphql-java KnownDirectives rule only recognizes executable directive locations, so ignore
      // the error if we're inside a type definition
      return PsiTreeUtil.getParentOfType(element, GraphQLTypeSystemDefinition::class.java) != null
    }

    return false
  }

  private fun intersectsInjectedFragments(element: PsiElement?): Boolean {
    val errorContext = if (element is GraphQLIdentifier) element.parent else element
    return errorContext != null && InjectedLanguageUtil.isInInjectedLanguagePrefixSuffix(errorContext)
  }

}
